clippy_lints/src/arbitrary_source_item_ordering.rs

```diff
@@ -0,0 +1,421 @@
+use clippy_config::types::{
+    SourceItemOrderingCategory, SourceItemOrderingModuleItemGroupings, SourceItemOrderingModuleItemKind,
+    SourceItemOrderingTraitAssocItemKind, SourceItemOrderingTraitAssocItemKinds,
+};
+use clippy_config::Conf;
+use clippy_utils::diagnostics::span_lint_and_note;
+use rustc_hir::{
+    AssocItemKind, FieldDef, HirId, ImplItemRef, IsAuto, Item, ItemKind, Mod, TraitItemRef, UseKind, Variant,
+    VariantData,
+};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::impl_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Confirms that items are sorted in source files as per configuration.
+    ///
+    /// ### Why restrict this?
+    ///
+    /// Keeping a consistent ordering throughout the codebase helps with working
+    /// as a team, and possibly improves maintainability of the codebase. The
+    /// idea is that by defining a consistent and enforceable rule for how
+    /// source files are structured, less time will be wasted during reviews on
+    /// a topic that is (under most circumstances) not relevant to the logic
+    /// implemented in the code. Sometimes this will be referred to as
+    /// "bike-shedding".
+    ///
+    /// Keep in mind, that ordering source code alphabetically can lead to
```

Comment:
This should be under a `# Known problems` header.

We also might want to add that the lint doesn't supply suggestions because it's quite hard to produce them in a correct way.

Reply:
Moved it down to the (now renamed) chapter "Known Problems". Is there a standard way to write this?

Ctrl+F "# Known" yields different ways to write this: "Known problems", "Known Problems", "Known issues"